openSearchConnection.go

```diff
@@ -0,0 +1,300 @@
+package lane
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/opensearch-project/opensearch-go/v3"
+	"github.com/opensearch-project/opensearch-go/v3/opensearchapi"
+)
+
+// Struct representing a connection to OpenSearch
+type openSearchConnection struct {
+	client             *opensearchapi.Client
+	clientMu           sync.Mutex
+	mu                 sync.Mutex
+	logBuffer          []*OslMessage
+	stopCh             chan *sync.WaitGroup
+	wakeCh             chan struct{}
+	clientCh           chan *sync.WaitGroup
+	refCount           int
+	config             *OslConfig
+	emergencyFn        OslEmergencyFn
+	ctx                context.Context
+	cancelFn           context.CancelFunc
+	messagesQueued     int
+	messagesSent       int
+	messagesSentFailed int
+}
+
+func (osc *openSearchConnection) connect(config *OslConfig) (err error) {
+
+	var client *opensearchapi.Client
+
+	if config == nil || !config.offline {
+		var wg sync.WaitGroup
+		wg.Add(1)
+		osc.clientCh <- &wg
+		wg.Wait()
+	}
+
+	osc.clientMu.Lock()
+
+	if config == nil {
+		config = &OslConfig{offline: true}
+	}
+
+	if !config.offline {
+		client, err = newOpenSearchClient(config.OpenSearchUrl, config.OpenSearchPort, config.OpenSearchUser, config.OpenSearchPass, config.OpenSearchTransport)
+		if err != nil {
+			return
+		}
+	}
+
+	osc.client = client
+	osc.config = config
+
+	if config.LogThreshold == 0 {
+		osc.config.LogThreshold = OslDefaultLogThreshold
+	}
+	if config.MaxBufferSize == 0 {
+		osc.config.MaxBufferSize = OslDefaultMaxBufferSize
+	}
+	if config.BackoffInterval == 0 {
+		osc.config.BackoffInterval = OslDefaultBackoffInterval
+	}
+	if config.BackOffLimit == 0 {
+		osc.config.BackOffLimit = OslDefaultBackOffLimit
+	}
+	osc.clientMu.Unlock()
+
+	return
+
+}
+
+func (osc *openSearchConnection) reconnect(config *OslConfig) (err error) {
+	return osc.connect(config)
+}
+
+func (osc *openSearchConnection) processConnection() {
+	backoffDuration := osc.config.BackoffInterval
+
+	for {
+		select {
+		case wg := <-osc.stopCh:
+			shouldStop := false
+			osc.mu.Lock()
+			osc.refCount--
+			shouldStop = (osc.refCount == 0)
+			osc.mu.Unlock()
+
+			if shouldStop {
+				go func() {
+					osc.mu.Lock()
+					if len(osc.logBuffer) > 0 {
```

Comment:
This isn't good. I had you change this to take ownership of the log buffer, instantly replace it so the client can begin logging more. That change has been reverted. Now it is back to the buggy behavior where it blocks on the I/O which can be indefinite, causing unexpected blockage in the client's app.